src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -108,12 +95,54 @@ public ODataBatchReaderState State
                 return this.batchReaderState;
             }
 
-            private set
+            protected set
             {
                 this.batchReaderState = value;
             }
         }
 
+        /// <summary>
+        /// >The batch-specific URL converter that stores the content IDs found in a changeset and supports resolving cross-referencing URLs.
+        /// </summary>
+        internal ODataBatchPayloadUriConverter PayloadUriConverter

```

Comment:
>internal [](start = 8, length = 8)

This should be protected, not internal.  If it's needed by a derived type, it shouldn't matter whether the derived type is in this assembly or a different assembly.

Reply:
the PayloadUriConverter is used in ReadImplementation() in base class, and also used in CreateOperationRequestMessageImplementation and CreateOperationResponseMessageImplementation, which are implemented in derived classes (as in "this.PayloadUriConverter.ContainsContentId(contentId)". I think by design it cannot be private? Let me know if I misunderstood. 

---
In reply to: [141139803](https://github.com/OData/odata.net/pull/960#discussion_r141139803) [](ancestors = 141139803)

Reply:
When we refactor, ReadImplementation should both validate that the ContentId is not already in the PayloadUriConverter and then add it, so we don't need to reference the PayloadUriConverter in CreateOperationResponseMessage.

---
In reply to: [141763790](https://github.com/OData/odata.net/pull/960#discussion_r141763790) [](ancestors = 141763790,141139803)

Reply:
It is also used for CreateOperation[Request|Response]MessageImplementation methods. To support the private encapsulation in base class, the following wrappers (protected) are added:
BuildOperationRequestMeassage
BuildOperationResponseMessage
CreateOperationRequestUriWrapper
AddToPayloadUriConverter
RememberContentIdHeader